src/rslice.rs

```diff
@@ -0,0 +1,184 @@
+use libc::{c_char, size_t};
+use std::convert::{TryFrom, TryInto};
+use std::marker::PhantomData;
+
+/// A read-only view on a Rust byte slice.
+///
+/// This is used to pass data from crustls to callback functions provided
+/// by the user of the API.
+/// `len` indicates the number of bytes than can be safely read.
+///
+/// The memory exposed is available as specified by the function
+/// using this in its signature. For instance, when this is a parameter to a
+/// callback, the lifetime will usually be the duration of the callback.
+/// Functions that receive one of these must not retain copies of any of the
+/// pointers beyond the allowed lifetime.
+#[repr(C)]
+pub struct rustls_slice_bytes<'a> {
+    pub data: *const u8,
+    pub len: size_t,
+    phantom: PhantomData<&'a [u8]>,
+}
+
+impl<'a> From<&'a [u8]> for rustls_slice_bytes<'a> {
+    fn from(s: &[u8]) -> Self {
+        rustls_slice_bytes {
+            data: s.as_ptr() as *const u8,
+            len: s.len() as size_t,
+            phantom: PhantomData,
+        }
+    }
+}
+
+/// An owned Vec<rustls_slice_bytes>, where the inner slices have lifetime 'a.
+/// If we want to share a view of a `Vec<Vec<_>>` with C, we can't do it
+/// directly because `Vec` is not #[repr(C)]. So we build a new Vec containing
+/// rustls_slice_bytes<'a> as an owned object. We can then expose slices of
+/// that Vec to C.
+pub(crate) struct VecSliceBytes<'a>(Vec<rustls_slice_bytes<'a>>);
```

Comment:
Could this be a type alias `type VecSliceBytes = Vec<rustls_slice_bytes<'a>>;` and save us the enclosing envelope struct?

Reply:
Unfortunately you can't implement traits on type aliases. That's one of the main reasons for using the newtype pattern: https://doc.rust-lang.org/book/ch19-03-advanced-traits.html#using-the-newtype-pattern-to-implement-external-traits-on-external-types